### ISSUE TITLE
CSS Changes

### ISSUE THREAD

**Issue by daniel-law** [EldinZenderink/LittleWeeb#10]
Hey, really loving the idea behind this program. In awe of your backend writing skills honestly. If you wouldn't mind I'd love to style the frontend and already have some concepts that I've worked on in the last few months. If I could, how and where would I need to go to spruce up the CSS?

**Comment by EldinZenderink** [comment]
Please do!  (though, it  isn't as simple as editing a css file:)

The interface is build using angular (5)/typescript, it generates normal javascript & html files. The way this interface is build up is by using a css framework library from semantic-ui. So the look and all the objects placed on the screen is actually part of html and not only css. So, unfortunately, for now, creating a whole new interface would mean writing a whole new interface, (which can be done!). You would need some javascript knowledge at least to write the communication with the backend through the api, which a wiki with explantion can be found here: https://github.com/EldinZenderink/LittleWeeb/wiki/BackEnd-API---How-to-use.

As long as the application is running (I would suggest using the server only variant for developing), you could test out your interface. If you want it to run with your interface, within the directory where the executable is, is also a map called GUI, in here, you can put your interface. Do name your front / main / home page as index.html, because that's the html file that the application will show you on startup. 

So TL;DR: 

I don't mind it at all, go ahead :D. I might add a wiki with all the custom interfaces made ;D. It's a bit difficult to do now though, as you can't go straight into css.

**Comment by daniel-law** [comment]
Hmm, it should be possible to theme the current interface rather than completely writing a new one I think. But it depends really how far you'd want to change the look i.e. the sidebar could moved to a topbar instead - then it'd be worth rewriting the current interface imo.

Did get the project to run last night (cloned master, grabbed the SimpleIRCLib from git & ran the solution in /Backend/Littleweeb.sln - this is the one to work on, right? Only problem is I can only view the actual interface in a seperate browser and not the actual program - guessing that's an issue with specific path name references. This the one to work on?

**Comment by EldinZenderink** [comment]
The /Backend/Littleweeb.sln is indeed the program to work on if you wanted to change the backend functionality. 

But in your case, you want to work on the frontend. This is how and why:

The reason why it isn't showing you the interface is because, when the executable is located in the Debug folder (where it's being build), it tries to load localhost:4200 through the winform cefsharp browser, which is the web address angular cli web host uses when trying to edit the interface,  so it would make developing the interface easier (for me atleast). Hence there is a separate directory for the interface on this github. You need [angular-cli](https://github.com/angular/angular-cli) to actually compile the interface. What you need to do to actually edit the interface code is as follows:

1. Install [angular-cli](https://github.com/angular/angular-cli).
2. (Preferably) Install [Visual Studio Code](https://code.visualstudio.com/)
3. Go to the /LittleWeebFrontend directory
4. Open the terminal command 
5. Right click while pressing shift within that directory
6. Click on Open Powershell here / Open Command window here (something like that)
7. type ng serve

If you then open the directory from step 3 within visual code, you can now edit the typescript code & css.

In your browser you can check the result by browsing to localhost:4200. Which will also be opened when you build the LittleWeeb through visual studio!

I do strongly recommend learning TypeScript and such for frontend work. Since the entire interface is written using TypeScript.

Also do note, I might forget a few steps now, but this is, out of my head, what needs to be done to be able to edit the interface.

**Comment by daniel-law** [comment]
Aight, thanks for those instructions, just wanted to see if I could get the backend to work too as well (might need to look in there sometime). I've tried to go through what you said and I'm getting errors on ng serve - seems I need a list of NPM pills to pop (angular2-toaster, rxjs, etc.). If you could tell me which ones I need that'd be great 😃

**Comment by EldinZenderink** [comment]
xD im so sorry for the inconvenience, It never occurred to me that I needed to keep track of those things (I am a typescript/angular noob after all). I guess I could take a quick look:
What you mentioned is correct so far:

- [angular2-toaster](https://www.npmjs.com/package/angular2-toaster)
- [rxjs](https://www.npmjs.com/package/rxjs)

Actually... after reading through my code, I think only these two need to be installed.

**Comment by daniel-law** [comment]
Aha, you're not. Anyway, after reading through the dependency errors I came up with the following list:

- angular-cli
- angular-devkit/schematics@0.0.46
- angular/common@^4.0.0
- angular/compiler@^4.0.0
- angular/compiler@^4.0.0
- rxjs@^5.0.0-beta.11
- zone.js@^0.8.4
- angular2-toaster

But the only problem is I still can't compile due to some sort of version mismatch (hoping you can actually decipher this for me...)
```
ERROR in Error: Metadata version mismatch for module C:/Users/Daniel/node_modules/@angular/core/core.d.ts, found version 4, expected 3, resolving symbol ToasterModule in C:/Users/Daniel/node_modules/angular2-toaster/src/toaster.module.d.ts, resolving symbol ToasterModule in C:/Users/Daniel/node_modules/angular2-toaster/src/toaster.module.d.ts
    at syntaxError (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@angular\compiler\bundles\compiler.umd.js:1729:34)
    at simplifyInContext (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@angular\compiler\bundles\compiler.umd.js:24979:23)
    at StaticReflector.simplify (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@angular\compiler\bundles\compiler.umd.js:24991:13)
    at StaticReflector.annotations (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@angular\compiler\bundles\compiler.umd.js:24418:41)
    at _getNgModuleMetadata (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@angular\compiler-cli\src\ngtools_impl.js:138:31)
    at _extractLazyRoutesFromStaticModule (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@angular\compiler-cli\src\ngtools_impl.js:109:26)
    at C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@angular\compiler-cli\src\ngtools_impl.js:129:27
    at Array.reduce (<anonymous>)
    at _extractLazyRoutesFromStaticModule (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@angular\compiler-cli\src\ngtools_impl.js:128:10)
    at Object.listLazyRoutesOfModule (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@angular\compiler-cli\src\ngtools_impl.js:53:22)
    at Function.NgTools_InternalApi_NG_2.listLazyRoutes (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@angular\compiler-cli\src\ngtools_api.js:91:39)
    at AotPlugin._getLazyRoutesFromNgtools (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@ngtools\webpack\src\plugin.js:212:44)
    at _donePromise.Promise.resolve.then.then.then.then.then (C:\Users\Daniel\Downloads\LittleWeeb-master\LittleWeeb-master\LittleWeebFrontend\node_modules\@ngtools\webpack\src\plugin.js:448:24)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)

webpack: Failed to compile.
```

**Comment by daniel-law** [comment]
Phew. Got it to compile with a bunch more packages. Only problem is that I'm still seeing a blank page in the program (localhost:4200).

**Comment by EldinZenderink** [comment]
Sry for the lack of response, It got real busy IRL ;X. Did you get it to work? If not, could you try (if compiling is succeeding) to use the command ng build instead of serve, and in the dist folder created by that, try to launch the index.html? (Using a webserver, for easiness: https://www.npmjs.com/package/http-server) (After installing, Go to the dist directory, right click in the directory while pressing shift-> Open command window/powershell window here -> type http-server, and then use the url shown in the command /powershell window )

By any chance, is there an error message in the Developer Tools  -> console window(if you use chrome) ?

**Comment by daniel-law** [comment]
I assume I did something wrong in the past. I got my desktop and managed to set it up using my set of instructions I put together with some more modifications. Get a working interface on debug and everything! Will get to work on custom styling - assuming white/blue is the color scheme you wanna go for?

Also, could you give me access to the Wiki - think it'd be useful to put my instructions there just in case anyone else wants to setup a dev environment in the future.

**Comment by EldinZenderink** [comment]
I feel stupid for asking... but how on earth do you grant permission for others to edit / contribute to the wiki? Do I need to add you as collaborator? Not that I have anything against it, but it feels kinda weird just allowing someone full access to the git :X.

**Comment by daniel-law** [comment]
I'm actually not sure if there is a way to add someone to the wiki at all. No worries, proposed a file change ([https://github.com/EldinZenderink/LittleWeeb/pull/11](https://github.com/EldinZenderink/LittleWeeb/pull/11)) to the README.md in the Frontend directory which gives instructions on how to setup.

Have made some slight changes already to the actual Frontend and fixed the ep 1/button issue in issue 9, up to you if you wanna add me as a collaborator or not. Can always do a pull request for the changes anyways ofc :) Added a screenshot to show you what I've changed 🖌 . 

![capture](https://user-images.githubusercontent.com/7337672/35465707-e4f27b0c-02f5-11e8-9d48-bab273160da4.jpg)

**Comment by EldinZenderink** [comment]
I merged it :D, you seem like a nice person, so of course I will add you as collaborator, (And I really need someone who can help with this kind of stuff 👍 )

**Comment by daniel-law** [comment]
Thank you, glad to help out :) Happy with my changes so far, what do you think though (buttons do go blue when you hover over them)?

![capture](https://user-images.githubusercontent.com/7337672/35476337-1b421d4c-03a6-11e8-9937-abb15ef09c87.JPG)

**Comment by EldinZenderink** [comment]
I like them :D, I have an update coming up as well (going to add a favorite page ;D). But that will give us a problem. When starting to create the interface... I didn't really think about letting others change the style. So some styling (I know ultra bad practice) is done directly in html, and classes for buttons define the colors as well. So... when I am going to push this update... either your or my work will be lost. To counter issues like this in the future, I will make it so that all the styling per component is done within their own css file. So that changing the component doesn't directly change individual styling of each object on screen. I hope sincerely that you didn't spend to much time on changing the interface as it is now, because unfortunately after the next update, it has to change again. (Btw, I updated everything related to the interface such as all the components (its now angular5-toaster instead of 2, and all the official components will be the latest version as well). 

I am sorry for these unfortunate changes (I am really not used to making things modular and changeable, it's a whole new dimension to think about. ;X).

**Comment by daniel-law** [comment]
Yup, did notice all the inline styling. I moved some of the styles (the buttons, menu and some padding) to the styles.css file which seems to be referenced automatically in all of the files :D Looking forward to seeing how you modify the design so I'll hold off pushing for now.

Also, just my opinion but I think it might be a good idea to move away from sematicui, it's not too bad but seems hard to work with from what I've encountered so far. For example the menu on the sidebar means the content on the right isn't centered properly. Can always move any styles we want now or in future to our own styles file.

**Comment by EldinZenderink** [comment]
Yeah, I was thinking about switching to bootstrap due to the great integration with angular. But it would take me at least days, and that's time I unfortunately do not have. I might reserve that as a TODO item for version v0.4.x . 

The very early version of littleweeb was written in plain jQuery, and since I didn't have much experience with frameworks in general in combination with angular / typescript I just went ahead and ported what I wrote in jQuery  /  html to Angular 2 (at the time). That included the semantic ui framework. Lately though I have done some side projects using bootstrap and I am really liking the simplicity of using that. (I am lazy after-all xD ). I am still not sure if giving every component it's own styling sheet is a good thing or not. I have the issue of overthinking and doing things. At the moment I am giving almost every component in html it's own class, which somewhat feels unnecessary but at the same time I could think up scenarios where someone would want to change the size of like one particular div or something ;).

**Comment by daniel-law** [comment]
It honestly depends I guess. The one thing I hate about having a CSS framework is that it's really bulky in the first place and you have to work around any pre-existing styles that you want to remove/want slightly modified.

Don't give every single component it's own styling sheet (assume you mean separate file). I'd break them down component by component in styles.css with a Global at the top for any recurring elements (like buttons). Actually what I did when modifying the menu. I overthink things all the time especially when writing code - but even if you do, everything can be simplified further and made to work better in the future, so no worries.

It is a lot of work, but arguably worth it in the end imo.

**Comment by EldinZenderink** [comment]
Hmmm might be my  noobiness about angular, but every component can have its own custom css file, like this:
![image](https://user-images.githubusercontent.com/12187560/35483699-808b7398-0445-11e8-9ddc-95b5ce32e5aa.png)

I personally thought it would be nice if you could per view/component change the css per like html element, but its done so that the original/default styling/class uses is  still there, like this:

![image](https://user-images.githubusercontent.com/12187560/35483706-8a0af740-0445-11e8-9f0a-ddf26ccb6f8a.png)

**Comment by daniel-law** [comment]
No, there's a global css file that's being referenced already - the /src/styles.css. Putting everything in that would be a good idea. By the way, probably a better idea to name things like this:

![naming](https://user-images.githubusercontent.com/7337672/35483843-ac9f31b0-043e-11e8-999d-9e5a693dc7eb.JPG)

**Comment by EldinZenderink** [comment]
Ah that does seem more logical. X) I was thinking of doing everything from bottom up, your way is more like top down ;). Welp, thank god I didn't do much yet.

**Comment by daniel-law** [comment]
The simpler, the better, especially with CSS ;)

**Comment by daniel-law** [comment]
```
.btn {
    background-color: transparent;
    cursor: pointer;
    display: inline-block;
    min-height: 1em;
    outline: none;
    border: none;
    vertical-align: baseline;
    font-family: 'Lato', 'Helvetica Neue', Arial, Helvetica, sans-serif;
    margin: 0em 0.1em 0em 0em;
    padding: 0.78571429em 1.5em 0.78571429em;
    margin: 0em 0.5em 0em 0em;
    text-transform: none;
    font-weight: bold;
    line-height: 1em;
    font-style: normal;
    text-align: center;
    text-decoration: none;
    border-radius: 0.28571429rem;
    transition: opacity 0.1s ease, background-color 0.1s ease, color 0.1s ease, box-shadow 0.1s ease, background 0.1s ease;
}

.grey {
    box-shadow: 0px 0px 0px 2px grey inset !important;
    color: grey;
}

.grey:hover {
    box-shadow: none !important;
    color: rgb(0, 162, 255);
    box-shadow: 0px 0px 0px 2px rgb(0, 162, 255) inset !important;
}

.blue {
    color: rgb(0, 162, 255);
    box-shadow: 0px 0px 0px 2px rgb(0, 162, 255) inset !important;
}
```

**Comment by EldinZenderink** [comment]
thanks :D, will add them, with your name above it ofcourse :D. After this update has finished you may push commits as well. Another Idea that's been passing through my mind: Should I make a separate repo for the interface (to keep the backend separate, as that one will more or less be done by me alone)? Technically speaking I could even host the interface through git-pages, but there might be some issues talking to the backend when using that. But not impossible to my believe. It would also mean that users won't have to re-download the whole application if only a front end has changed.

**Comment by daniel-law** [comment]
Tbh I think it'd be more effort that reward. For frontend it doesn't really matter when they're implemented as long as they work properly with the backend. Would be better just to push out a new version when enough things have been changed.

Think we need a way of communicating other than through an issue, already getting way too long xD Email me your discord details/other similar platform you use (my contact details are on https://github.com/daniel-law) and I'll add you so we can inform each other on what/where we're up to ;)

**Comment by EldinZenderink** [comment]
I could just create a discord group especially for littleweeb development ;), here it is: https://discord.gg/CRyc6 (link expires in 30 minutes)
